components/tikv_util/src/config.rs

```diff
@@ -1569,18 +1570,66 @@ impl RaftDataStateMachine {
         }
     }
 
+    // Remove all files and directories under `remove_path` except `retain_path`.
+    fn must_remove_except(remove_path: &Path, retain_path: &Path) {
+        if !remove_path.exists() {
+            info!("Path not exists"; "path" => %remove_path.display());
+            return;
+        }
+        if !remove_path.is_dir() {
+            info!("Path is not a directory, so remove directly"; "path" => %remove_path.display());
+            Self::must_remove(remove_path);
+            return;
+        }
+        if !retain_path.starts_with(remove_path) {
+            info!("Retain path is not under the remove path, so remove directly"; "Retain path" => %retain_path.display(), "remove path" => %remove_path.display());
+            Self::must_remove(remove_path);
+            return;
+        }
+
+        for entry in fs::read_dir(remove_path).unwrap() {
+            let sub_path = entry.unwrap().path();
+            if sub_path != retain_path {
+                Self::must_remove(&sub_path);
+            }
+        }
+    }
+
     fn must_rename_dir(from: &Path, to: &Path) {
         fs::rename(from, to).unwrap();
         let mut dir = to.to_path_buf();
         assert!(dir.pop());
         Self::sync_dir(&dir);
     }
 
-    fn data_exists(path: &Path) -> bool {
-        if !path.exists() || !path.is_dir() {
-            return false;
+    pub fn raftengine_exist(raftengine_path: &Path) -> bool {
+        if raftengine_path.is_dir() {
+            for entry in raftengine_path.read_dir().unwrap() {
+                match entry {
+                    Ok(entry) => {
+                        let path = entry.path();
+                        if path.is_file() && path.extension() == Some("raftlog".as_ref()) {
+                            return true;
+                        }
+                    }
+                    Err(e) => {
+                        warn!("Scan raft engine dir failed"; "path" => %raftengine_path.display(), "err" => ?e)
+                    }
+                };
+            }
         }
-        fs::read_dir(path).unwrap().next().is_some()
+        false
+    }
+
+    pub fn raftdb_exist(raftdb_path: &Path) -> bool {
+        if raftdb_path.is_dir() && raftdb_path.join("CURRENT").exists() {
+            return true;
+        }
+        false
+    }
+
+    pub fn raft_data_exists(path: &Path) -> bool {
+        Self::raftengine_exist(path) || Self::raftdb_exist(path)
```

Comment:
IMO, it's redundant to rename `data_exists` to `raft_data_exists`, as the naming of this class already contains this.
```suggestion
    pub fn data_exists(path: &Path) -> bool {
        Self::raftengine_exists(path) || Self::raftdb_exists(path)
```